File: zig/test/regex_pathological.zig
// Discovery test: pathological regex inputs run against the port's re_* API.
// Goal is to surface failures across ports, not to assert behaviour.
// Panel is the same in every port (see REGEX.md).
//
// Zig's public regex surface currently exposes only re_compile/re_test/re_escape
// (see src/struct.zig). The find/replace/find_all cases below mark themselves
// as N/A — that absence is itself part of the discovery.

const std = @import("std");
const voxgig_struct = @import("voxgig-struct");

fn record_test(label: []const u8, ok: bool, ms: f64, value: anytype) void {
    const T = @TypeOf(value);
    const writer = std.io.getStdOut().writer();
    if (ok) {
        if (T == bool) {
            writer.print("[regex-discovery] {s} | {d:.2}ms | OK | {}\n", .{ label, ms, value }) catch {};
        } else {
            writer.print("[regex-discovery] {s} | {d:.2}ms | OK | {any}\n", .{ label, ms, value }) catch {};
        }
    } else {
        writer.print("[regex-discovery] {s} | {d:.2}ms | ERR | compile or run failed\n", .{ label, ms }) catch {};
    }
}

test "regex pathological discovery" {
    var buf: [4096]u8 = undefined;
    var fba = std.heap.FixedBufferAllocator.init(&buf);
    const alloc = fba.allocator();

    const a22 = try alloc.alloc(u8, 22);
    @memset(a22, 'a');
    const p1_in = try std.fmt.allocPrint(alloc, "{s}!", .{a22});

    var nest_buf: [120]u8 = undefined;
    var pos: usize = 0;
    while (pos < 40) : (pos += 1) nest_buf[pos] = '(';
    nest_buf[pos] = 'a';
    pos += 1;
    var i: usize = 0;
    while (i < 40) : (i += 1) {
        nest_buf[pos] = ')';
        pos += 1;
    }
    const nest40 = nest_buf[0..pos];

    // P1
    var t0 = std.time.nanoTimestamp();
    const b1 = voxgig_struct.re_test("^(a+)+$", p1_in);
    var ms = @as(f64, @floatFromInt(std.time.nanoTimestamp() - t0)) / 1e6;
    record_test("P1_redos_nested_plus", true, ms, b1);

    // P2
    t0 = std.time.nanoTimestamp();
    const b2 = voxgig_struct.re_test("^(a|aa)+$", p1_in);
    ms = @as(f64, @floatFromInt(std.time.nanoTimestamp() - t0)) / 1e6;
    record_test("P2_redos_alt_overlap", true, ms, b2);

    // P3, P4, P5, P10 — replace/find/find_all not in zig public surface.
    const writer = std.io.getStdOut().writer();
    try writer.print("[regex-discovery] P3_empty_repeat_replace | -.--ms | N/A | re_replace not exposed\n", .{});
    try writer.print("[regex-discovery] P4_unicode_replace_dot | -.--ms | N/A | re_replace not exposed\n", .{});
    try writer.print("[regex-discovery] P5_unicode_find_codepoint | -.--ms | N/A | re_find not exposed\n", .{});

    // P6
    t0 = std.time.nanoTimestamp();
    const b6 = voxgig_struct.re_test(nest40, "a");
    ms = @as(f64, @floatFromInt(std.time.nanoTimestamp() - t0)) / 1e6;
    record_test("P6_deep_nesting_compile", true, ms, b6);

    // P7
    t0 = std.time.nanoTimestamp();
    const b7 = voxgig_struct.re_test("^a{0,10000}b$", "aaaaaaaaaab");
    ms = @as(f64, @floatFromInt(std.time.nanoTimestamp() - t0)) / 1e6;
    record_test("P7_big_bounded_quantifier", true, ms, b7);

    // P8
    t0 = std.time.nanoTimestamp();
    const p8 = voxgig_struct.re_compile("[abc");
    ms = @as(f64, @floatFromInt(std.time.nanoTimestamp() - t0)) / 1e6;
    if (p8 == null) {
        try writer.print("[regex-discovery] P8_invalid_pattern | {d:.2}ms | ERR | compile returned null\n", .{ms});
    } else {
        try writer.print("[regex-discovery] P8_invalid_pattern | {d:.2}ms | OK | \"compiled\"\n", .{ms});
    }

    // P9
    t0 = std.time.nanoTimestamp();
    const b9 = voxgig_struct.re_test("^(a+)\\1$", "aaaa");
    ms = @as(f64, @floatFromInt(std.time.nanoTimestamp() - t0)) / 1e6;
    record_test("P9_backref_re2_forbidden", true, ms, b9);

    try writer.print("[regex-discovery] P10_find_all_zero_width | -.--ms | N/A | re_find_all not exposed\n", .{});
}

File: lua/test/regex_pathological.lua
-- Discovery test: pathological regex inputs run against the port's re_* API.
-- Goal is to surface failures across ports, not to assert behaviour.
-- Panel is the same in every port (see REGEX.md).
--
-- RUN: lua test/regex_pathological.lua

package.path = "../src/?.lua;./src/?.lua;" .. (package.path or "")
local re = require("regex")

local function json_str(s)
  return '"' .. tostring(s):gsub('"', '\\"') .. '"'
end

local function json_table(t)
  local parts = {}
  for _, v in ipairs(t) do
    if type(v) == "table" then
      parts[#parts + 1] = json_table(v)
    elseif type(v) == "string" then
      parts[#parts + 1] = json_str(v)
    else
      parts[#parts + 1] = tostring(v)
    end
  end
  return "[" .. table.concat(parts, ",") .. "]"
end

local function render(r)
  local t = type(r)
  if t == "nil" then return "null"
  elseif t == "boolean" then return tostring(r)
  elseif t == "string" then return json_str(r)
  elseif t == "table" then return json_table(r)
  else return tostring(r) end
end

local function record(label, fn)
  local t0 = os.clock()
  local ok, r = pcall(fn)
  local ms = (os.clock() - t0) * 1000.0
  local outcome
  if ok then
    outcome = "OK | " .. render(r)
  else
    outcome = "ERR | " .. tostring(r)
  end
  io.write(string.format("[regex-discovery] %s | %.2fms | %s\n", label, ms, outcome))
end

local a22 = string.rep("a", 22)
local nest40 = string.rep("(", 40) .. "a" .. string.rep(")", 40)

record("P1_redos_nested_plus",      function() return re.re_test("^(a+)+$", a22 .. "!") end)
record("P2_redos_alt_overlap",      function() return re.re_test("^(a|aa)+$", a22 .. "!") end)
record("P3_empty_repeat_replace",   function() return re.re_replace("a*", "abc", "X") end)
record("P4_unicode_replace_dot",    function() return re.re_replace("\\.", "café.au.lait", "/") end)
record("P5_unicode_find_codepoint", function() return re.re_find("é", "café au lait") end)
record("P6_deep_nesting_compile",   function() return re.re_test(nest40, "a") end)
record("P7_big_bounded_quantifier", function() return re.re_test("^a{0,10000}b$", string.rep("a", 10) .. "b") end)
record("P8_invalid_pattern",        function() return re.re_compile("[abc") end)
record("P9_backref_re2_forbidden",  function() return re.re_test("^(a+)\\1$", "aaaa") end)
record("P10_find_all_zero_width",   function() return re.re_find_all("a*", "bbb") end)
